### PR TITLE
MicrosoftGraphUser add manager information

### PR DESCRIPTION
**Pull request by JT-NL**
## Status
- [X] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
Addition of the commands to the existing Microsoft Graph User integration. Adds a command to retrieve the manager information for a user and to assign a manager to a user. Documentation and test playbook is also updated to include the direct reports commands, as this was missing.

## Minimum version of Demisto
- Unchanged from existing integration

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No
     Although the (existing) direct reports commands does not follow the convention of the other graph user commands, this was not updated to maintain backward compatibility

## Must have
- [X] Tests
- [X] Documentation

### CODE REVIEW

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/README.md [hunk 1]
```diff
@@ -1119,5 +1122,425 @@ For more details about the authentication used in this integration, see <a href=
 </div>
 <p> </p>
 <div class="cl-preview-section">
-<p><img src="https://user-images.githubusercontent.com/37335599/56651529-e92f6800-6691-11e9-8bd1-8779a59cf51d.png" alt="Screen Shot 2019-04-24 at 12 13 05"></p>
```
Comment (JT-NL) [inline Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/README.md:1122]: This screenshot [https://user-images.githubusercontent.com/37335599/56651529-e92f6800-6691-11e9-8bd1-8779a59cf51d.png](https://user-images.githubusercontent.com/37335599/56651529-e92f6800-6691-11e9-8bd1-8779a59cf51d.png) was from the wildfire integration

**Comment by Itay4** [comment]
hi @JT-NL 
i see the status of the PR is marked as In Progress
can you please comment when it's ready for review ?

**Comment by JT-NL** [comment]
> hi @JT-NL
> i see the status of the PR is marked as In Progress
> can you please comment when it's ready for review ?

Hi @Itay4
The PR just passed all automated tests, so it is now ready for review

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py [hunk 2]
```diff
@@ -55,6 +55,7 @@ class MsGraphClient:
     def __init__(self, tenant_id, auth_id, enc_key, app_name, base_url, verify, proxy, self_deployed):
         self.ms_client = MicrosoftClient(tenant_id=tenant_id, auth_id=auth_id, enc_key=enc_key, app_name=app_name,
                                          base_url=base_url, verify=verify, proxy=proxy, self_deployed=self_deployed)
+        self.base_url = base_url
```
Comment (Itay4) [inline Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py:58]: don't think it's needed, you can remove

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py [hunk 3]
```diff
@@ -135,6 +136,26 @@ def get_direct_reports(self, user):
...
+        manager_data = self.ms_client.http_request(
+            method='GET',
+            url_suffix=f'users/{user}/manager')
+        manager_data.pop('@odata.context', None)
+        return manager_data
+
+    #  If successful, this method returns 204 No Content response code.
+    #  Using resp_type=text to avoid parsing error.
+    def assign_manager(self, user, manager):
+        url = self.base_url
```
Comment (Itay4) [inline Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py:149]: don't think it's needed, you can remove

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.yml [hunk 4]
```diff
@@ -437,9 +443,72 @@ script:
...
+      required: true
+      secret: false
+    deprecated: false
+    description: "Retrieves the properties from the manager of a user"
+    execution: false
+    name: msgraph-user-get-manager
+    outputs:
+    - contextPath: MSGraphUserManager.ID
+      description: User's ID.
+      type: Unknown
```
Comment (Itay4) [inline Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.yml:460]: please set the output types

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py [hunk 5]
```diff
@@ -264,6 +285,29 @@ def get_direct_reports_command(client: MsGraphClient, args: Dict):
...
+    }
+    return human_readable, outputs, manager_data
+
+
+def assign_manager_command(client: MsGraphClient, args: Dict):
+    user = args.get('user')
+    manager = args.get('manager')
+    client.assign_manager(user, manager)
+    human_readable = f'A manager was assigned to user "{user}". It might take several minutes for the changes ' \
+                     f'to take affect across all applications. '
```
Comment (Itay4) [inline Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py:307]:
```suggestion
                     'to take affect across all applications.'
```

File: Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml [hunk 6]
```diff
@@ -545,17 +532,231 @@ tasks:
...
+      script: '|||msgraph-user-get-manager'
+      type: regular
+      iscommand: true
+      brand: ""
+    nexttasks:
+      '#none#':
+      - "23"
+    scriptarguments:
+      user:
+        simple: ${GeneratedUUID}test@${inputs.baseurl}
```
Comment (Itay4) [inline Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml:664]: let's verify the expected outputs are returned

File: Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml [hunk 7]
```diff
@@ -545,17 +532,231 @@ tasks:
...
+      script: '|||msgraph-direct-reports'
+      type: regular
+      iscommand: true
+      brand: ""
+    nexttasks:
+      '#none#':
+      - "17"
+    scriptarguments:
+      user:
+        simple: ${GeneratedUUID}manager@${inputs.baseurl}
```
Comment (Itay4) [inline Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml:697]: let's verify the expected outputs are returned

**Comment by kirbles19** [comment]
@Itay4 Done.

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py (hunk 2 above)
Comment (JT-NL) [reply Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py:58]: Currently using this to get the base url required for the assign_manager request body. The body has to be formatted like this [example from the API docs](https://docs.microsoft.com/en-us/graph/api/user-post-manager?view=graph-rest-1.0&tabs=http)
`PUT https://graph.microsoft.com/v1.0/users/{id}/manager/$ref
Content-type: application/json
Content-length: xxx

{
  "@odata.id": "https://graph.microsoft.com/v1.0/users/{id}"
}`
Do you have a suggestion for constructing this body in another way?

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py (hunk 3 above)
Comment (JT-NL) [reply Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py:149]: See previous comment

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.yml (hunk 4 above)
Comment (JT-NL) [reply Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.yml]: This part was mostly copied from other commands ;)
I added the output types in the latest commit

File: Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml (hunk 6 above)
Comment (JT-NL) [reply Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml:664]: How do you want to do that exactly?
Just visually with Azure AD next to it and/or adding conditions with operator: isNotEmpty as used in auto generated test playbooks by the demisto-sdk generate-test-playbook command?

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py (hunk 2 above)
Comment (Itay4) [reply Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py:58]: it is already stored
look at https://github.com/demisto/content/blob/master/Packs/Base/Scripts/CommonServerPython/CommonServerPython.py#L3525

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py [hunk 8]
```diff
@@ -135,6 +136,26 @@ def get_direct_reports(self, user):
...
+            method='GET',
+            url_suffix=f'users/{user}/manager')
+        manager_data.pop('@odata.context', None)
+        return manager_data
+
+    #  If successful, this method returns 204 No Content response code.
+    #  Using resp_type=text to avoid parsing error.
+    def assign_manager(self, user, manager):
+        url = self.base_url
+        manager_ref = "{}users/{}".format(url, manager)
```
Comment (Itay4) [inline Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py:150]:
```suggestion
        manager_ref = "{}users/{}".format(self.base_url, manager)
```

File: Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml (hunk 6 above)
Comment (Itay4) [reply Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml:664]: you can use conditions to see the returned outputs are the expected ones (i.e. the manager is what you set)

you can use any condition/filter you prefer (`Equals` can check the manager returned is the one you set for example)

**Comment by ShahafBenYakir** [comment]
Hi there @JT-NL , are you planning on referring to the changes asked by @Itay4? If not we will close this PR in a couple of days.

**Comment by JT-NL** [comment]
> Hi there @JT-NL , are you planning on referring to the changes asked by @Itay4? If not we will close this PR in a couple of days.

Hi @ShahafBenYakir, I was quite busy, so I totally forgot about this PR.
I'll try to send an updated PR this weekend :)

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py (hunk 2 above)
Comment (JT-NL) [reply Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py]: Updated

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py (hunk 3 above)
Comment (JT-NL) [reply Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py]: Updated, removed self.base_url = base_url by referring to the existing variable

File: Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml (hunk 6 above)
Comment (JT-NL) [reply Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml]: Update the testing playbook for the manager related commands

**Comment by JT-NL** [comment]
@Itay4  and @ShahafBenYakir, I updated the pull request with the requested changes.
Can you review the changes?

PS: note that I synced with the master branch before the latest PR (see https://github.com/demisto/content/compare/master...JT-NL:master)

**Comment by Itay4** [comment]
> @Itay4 and @ShahafBenYakir, I updated the pull request with the requested changes.
> Can you review the changes?
> 
> PS note:
> 
> * that I synced with the master branch before the latest PR (see [master...JT-NL:master](https://github.com/demisto/content/compare/master...JT-NL:master))
> * there is a message "license/cla Pending — Contributor License Agreement is not signed yet.", however I don't have a possibility to sign (again) because I already signed that CLA

No worries about the CLA, just had to refresh it - fixed now

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.yml [hunk 9]
```diff
@@ -410,25 +410,90 @@ script:
...
+      required: true
+      secret: false
+    - default: false
+      description: User ID or userPrincipalName of the manager.
+      isArray: false
+      name: manager
+      required: true
+      secret: false
+    deprecated: false
+    description: "Assigns a manager to the specified user. Permissions: - User.ReadWrite (Delegated & Application)"
```
Comment (Itay4) [inline Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.yml:494]: @JT-NL are you sure both permissions, delegated and application are required? i think application is enough

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py (hunk 8 above)
Comment (JT-NL) [reply Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.py]: Updated, removed self.base_url = base_url by referring to the existing variable

File: Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.yml (hunk 9 above)
Comment (JT-NL) [reply Packs/MicrosoftGraphUser/Integrations/MicrosoftGraphUser/MicrosoftGraphUser.yml]: @Itay4  No, only one of them is [required](https://docs.microsoft.com/en-us/graph/api/user-post-manager?view=graph-rest-1.0&tabs=http). I updated the description to make that clear.

File: Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml [hunk 10]
```diff
@@ -310,22 +305,21 @@ tasks:
...
-    taskid: 5e97172f-3717-4fe5-8f4f-8beb05667f94
+    taskid: 5ace3b5a-c0b2-4221-8611-90a4ac358ed9
     type: title
     task:
-      elasticcommonfields: {}
-      id: 5e97172f-3717-4fe5-8f4f-8beb05667f94
+      id: 5ace3b5a-c0b2-4221-8611-90a4ac358ed9
       version: -1
-      name: Done!
+      name: Error!
```
Comment (Itay4) [inline Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml:313]: if you consider this an error, need to return an error entry, as currently the test will pass

Comment (JT-NL) [reply Packs/MicrosoftGraphUser/TestPlaybooks/Microsoft_Graph_User_test.yml]: @Itay4 I just added the two 'PrintErrorEntry' tasks to test playbook. 
Please not that this step is currently not mentioned on the dev [page](https://xsoar.pan.dev/docs/integrations/test-playbooks).

**Comment by Itay4** [comment]
@JT-NL looks great, thanks for the time and effort !
we appreciate your contribution and looking forward for the next one :wink:

**Review by Itay4 — APPROVED** [review]

**Pull request merged**